File: 0875-koko-eating-bananas/0875-koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int max = 0;

        for (int num : piles) max = Math.max(max, num);

        int l = 0;
        int r = max;
        int mid;

        while (l < r) {
            mid = l + (r - l) / 2;
            if (helper(mid, h, piles)) {
                r = mid;
            } else {
                l = mid + 1;
            }
        }
        return r;
    }

    private boolean helper(int rate, int h, int[] piles) {
        for (int i = 0; i < piles.length; i++) {
            h -= Math.ceil((double) piles[i] / (double) rate);
            if (h < 0) return false;
        }
        return true;
    }
}

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.cpp
class Solution {
bool isPossible(vector<int>& piles, int h, int speed) {
    int hours{};

    for (auto i{0uz}; i < piles.size(); ++i) {
        hours += piles[i] / speed;
        if (piles[i] % speed != 0) ++hours;
    }

    return hours <= h;
}
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int l{1};
        int r{*std::max_element(piles.begin(), piles.end())};

        while (l < r) {
            int mid = l + (r - l) / 2;

            if (isPossible(piles, h, mid)) {
                r = mid;
            } else {
                l = mid + 1;
            }
        }

        return r;
    }
};